autogen/agentchat/contrib/agent_builder.py

```diff
@@ -0,0 +1,450 @@
+        whether this task need programming to solve.
+    After that, user can call start() at a proper time to complete the task.
+    User can save the built agents' config by calling save(), and load the saved configs by load(), which can skip the
+        building process.
+    """
+
+    openai_server_name = "openai"
+    max_tokens = 945
+    max_agents = 5  # maximum number of agents build manager can create.
+
+    CODING_PROMPT = """Does the following task need programming (i.e., access external API or tool by coding) to solve,
+    or use program may help the following task become easier?
+
+    TASK: {task}
+
+    Hint:
+    # Answer only YES or NO.
+    """
+
+    AGENT_NAME_PROMPT = """To complete the following task, what positions/jobs should be set to maximize the efficiency?
+
+    TASK: {task}
+
+    Hint:
+    # Considering the effort, the position in this task should be no more then {max_agents}, less is better.
+    # Answer the name of those positions/jobs, separated by comma and use "_" instead of space. For example: Product_manager,Programmer
+    # Only return the list of positions.
+    """
+
+    AGENT_SYS_MSG_PROMPT = """Considering the following position and corresponding task:
+
+    TASK: {task}
+    POSITION: {position}
+
+    Modify the following position requirement, let it more suitable for the above task and position:
+
+    REQUIREMENT: {default_sys_msg}
+
+    Hint:
+    # The modified requirement should not contain the code interpreter skill.
+    # Coding skill is limited to Python.
+    # Your answer should omit the word "REQUIREMENT".
+    # Your should let them reply "TERMINATE" in the end when the task complete (user's need has been satisfied).
+    """
+
+    def __init__(
+        self,
+        config_path: Optional[str] = "OAI_CONFIG_LIST",
+        builder_model: Optional[str] = "gpt-4-1106-preview",
+        agent_model: Optional[str] = "gpt-4-1106-preview",
+        host: Optional[str] = "localhost",
+        endpoint_building_timeout: Optional[int] = 600,
+    ):
+        """
+        Args:
+            config_path: path of the OpenAI api configs.
+            builder_model: specify a model as the backbone of build manager.
+            host: endpoint host.
+            endpoint_building_timeout: timeout for building up an endpoint server.
+        """
+        self.host = host
+        self.builder_model = builder_model
+        self.agent_model = agent_model
+        self.config_path = config_path
+        self.endpoint_building_timeout = endpoint_building_timeout
+
+        self.user_proxy: autogen.UserProxyAgent = None
+        self.group_chat_manager_config: dict = None
+        self.initiate_agent_name: str = "User_console_and_Python_code_interpreter"
+        self.manager_system_message: str = "Group chat manager."
+        self.agent_configs: List[Dict] = []
+        self.coding: bool = None
```

Comment:
I think `coding` should be a build time parameter, not part of the builder specification. Ideally it would be great to have the builder's internal state reusable across multiple `build` calls.